File: Solutions/2-Add_Two_Numbers/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
    {
        var headNode = new ListNode(0);
        var currentNode = headNode;
        
        var x = l1;
        var y = l2;
        var carry = 0;
        
        while (x != null || y != null)
        {
            int a = (x != null) ? x.val : 0;
            int b = (y != null) ? y.val : 0;
            
            var sum = carry + a + b;      
            carry = sum / 10;
            currentNode = currentNode.next = new ListNode(sum % 10);
            
            if (x != null) { x = x.next; }
            if (y != null) { y = y.next; }
        }
        
        // Check if there is still a value to carry over after finishing the loop
        if (carry == 1)
        { currentNode.next = new ListNode(1); }
        
        return headNode.next;
    }
}

// Another
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(-1, null);
        ListNode first = res;
        int helper = 0;
        while(l1 != null || l2 != null){
            int currVal = ((l1?.val ?? 0) + (l2?.val ?? 0) + helper);
            res.val = currVal % 10;
                
            helper = currVal / 10;
            
            if(l1?.next != null || l2?.next != null || helper > 0){
                res.next = new ListNode(helper, null);
                res = res.next;
            }
            
            l1 = l1?.next;
            l2 = l2?.next;
        }
        
        return first;
    }
}

// Fastest
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        var temp = new ListNode(0);
        var dummy = temp;
        var carry=0;
        while(l1 != null || l2 != null || carry > 0)
        {
            var sum = 0;
            if(l1 != null)
            {
                sum += l1.val;
                l1 = l1.next;
            }
            
            if(l2 != null)
            {
                sum += l2.val;
                l2 = l2.next;
            }
            
            sum += carry;
            carry= 0;
            if(sum > 9)
            {
                carry = sum/10;
                sum = sum%10;
            }
            
            dummy.next = new ListNode(sum);
            dummy = dummy.next;
        }
        
        return temp.next;
    }
}

File: Solutions/2-Add_Two_Numbers/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy=new ListNode();
        ListNode* temp=dummy;
        int carry=0;
        while(l1!=NULL || l2!=NULL || carry){
            int sum=0;
            if(l1!=NULL){
                sum+=l1->val;
                l1=l1->next;
            }
            if(l2!=NULL){
                sum+=l2->val;
                l2=l2->next;
            }
            sum+=carry;
            carry=sum/10;
            ListNode* newnode=new ListNode(sum%10);
            temp->next=newnode;
            temp=temp->next;
        }
        return dummy->next;
    }
};

// Definitely faster than mine
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *ans = new ListNode(0);
        ListNode *ret = ans;
        if(!l1)
            return l2;
        if(!l2)
            return l1;
        int carry = 0;
        while(l1 && l2)
        {
            int v = l1->val + l2->val;
            v+=carry;
            ListNode *t = new ListNode(v%10);
            ans->next = t;
            carry = v/10;
            ans = ans->next;
            l1=l1->next;
            l2=l2->next;
        }   
        if(l1)
        {
            while(l1)
            {
                int v = l1->val;
                v+=carry;
                ListNode *t = new ListNode(v%10);
                ans->next = t;
                carry = v/10;
                ans = ans->next;
                l1=l1->next;
            }
        }
        else if(l2)
        {
            while(l2)
            {
                int v = l2->val;
                v+=carry;
                ListNode *t = new ListNode(v%10);
                ans->next = t;
                carry = v/10;
                ans = ans->next;
                l2=l2->next;
            }
        }
        if(carry)
        {
            ListNode *t = new ListNode(1);
            ans->next = t;
        }
        return ret->next;
    }
};


// Almost the fastest
void check(char& d, char& c) {
    if (d > '9') {
        d -= 10;
        c = 1;
    } else {
        c = 0;
    }
}
int main(int argc, char *argv[])
{
    ofstream fout("user.out");
    
    char c;
    
    vector<char> num;
    while (cin >> c && c == '[') {
        int length1 = 0;
        char val, c2;
        num.clear();
        
        do {
            val = getchar();
            c2 = getchar();
            ++length1;
            num.push_back(val);
        } while (c2 == ',');
        scanf(" %c", &c);
        c = 0;
        
        int length2 = 0;
        do {
            val = getchar();
            c2 = getchar();
            char d;
            if (length2 < length1) {
                d = num[length2] + val - '0' + c;
                check(d, c);
            } else {
                d = val + c;
                check(d, c);
            }
            fout << (length2 ? ',' : '[') << d;
            ++length2;
        } while (c2 == ',');
        while (length2 < length1) {
            char d = num[length2] + c;
            if (c) check(d, c);
            fout << ',' << d;
            ++length2;
        }
        if (c != 0) {
            fout << ",1";
        }
        fout << "]\n";
    }
}
#define main deleted_main
struct Solution {
    template<typename... Args>
    ListNode* addTwoNumbers(Args... args) {
        return {};
    }
};

// Fastest
ofstream ans("user.out");
vector<int> numbers;
char ch;
int main()
{
    while(cin >> ch)
    {
        int sum = 0, c = 0;
        numbers.clear();
        do
        {
            numbers.push_back(getchar() - '0');
            ch = getchar();
        } while(ch == ',');
        cin >> ch;
        do
        {
            ch = getchar();
            if(c < numbers.size()) sum += numbers[c];
            sum += ch - '0';
            ans << (c ? ',' : '[') << sum % 10;
            sum /= 10;
            c++;
            ch = getchar();
        } while(ch == ',');
        while(c < numbers.size())
        {
            sum += numbers[c];
            ans << ',' << sum % 10;
            sum /= 10;
            c++;
        }
        if(sum) ans << ",1";
        ans << "]" << endl;
    }
    
}
#define main deleted_main
struct Solution
{
    template<typename... l1>
    ListNode* addTwoNumbers(l1... l2)
    {
        return {};
    }
};